src/v/utils/tracing_mutex.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/semaphore.hh>
+
+#include <utility>
+
+/*
+ * This is a extension of mutex.h. The primary motivation for
+ * this class is to gather statistics on how ofter the mutex is
+ * taken (lock_counter) and the id of the top lock holder.
+ *
+ * Usage
+ * =====
+ *
+ *    ```
+ *    enum holder_id { tracing_mutex_example = 0 }
+ *
+ *    mutex m;
+ *    return m.with(holder_id::tracing_mutex_example, [] { ... });
+ *    ```
+ *
+ */
+
+class tracing_mutex;
+
+class locked_token {
+    tracing_mutex* _mutex;
+    ss::semaphore_units<> _unit;
+    bool _is_moved{false};
```

Comment:
Instead of using raw pointer and flag I'd suggest to use `std::optional<std::reference_wrapper<tracing_mutex>>`. You can cast it to bool to check if it contains value or not. And if the optional was moved it will return false, so you can check it in the d-tor.  Given that, you can get rid of explicit move c-tor. Simple `locked_token(locked_token&& token) = default` will generate correct c-tor.